test/schema-feature.test.js

```diff
@@ -1289,3 +1290,320 @@ test('setSchemaController per instance', t => {
 
   fastify.ready(err => { t.error(err) })
 })
+
+test('setSchemaController: Inherits correctly parent schemas with a customized validator instance', async t => {
+  t.plan(5)
+  const customAjv = new Ajv({ coerceTypes: false })
+  const server = Fastify()
+  const someSchema = {
+    $id: 'some',
+    type: 'array',
+    items: {
+      type: 'string'
+    }
+  }
+  const errorResponseSchema = {
+    $id: 'error_response',
+    type: 'object',
+    properties: {
+      statusCode: {
+        type: 'integer'
+      },
+      message: {
+        type: 'string'
+      }
+    }
+  }
+
+  // these schemas must be added before calling setSchemaController()
+  server.addSchema(someSchema)
+  server.addSchema(errorResponseSchema)
+
+  server.register((instance, _, done) => {
+    instance.setSchemaController({
+      compilersFactory: {
+        buildValidator: function (externalSchemas) {
+          const schemaKeys = Object.keys(externalSchemas)
+          t.equal(schemaKeys.length, 2, 'Contains same number of schemas')
+          t.hasStrict([someSchema, errorResponseSchema], Object.values(externalSchemas), 'Contains expected schemas')
+          for (const key of schemaKeys) {
+            if (customAjv.getSchema(key) == null) {
+              customAjv.addSchema(externalSchemas[key], key)
+            }
+          }
+          return function validatorCompiler ({ schema }) {
+            return customAjv.compile(schema)
+          }
+        }
+      }
+    })
+
+    instance.get(
+      '/',
+      {
+        schema: {
+          querystring: {
+            msg: {
+              $ref: 'some#'
+            }
+          },
+          response: {
+            '4xx': {
+              $ref: 'error_response#'
+            }
+          }
+        }
+      },
+      (req, reply) => {
+        reply.send({ noop: 'noop' })
+      }
+    )
+
+    done()
+  })
+
+  const res = await server.inject({
+    method: 'GET',
+    url: '/',
+    query: {
+      msg: ['string']
```

Comment:
this is less confusing. it is not immediate that an array within one element is used as a single entity


```suggestion
      msg: 'string'
```